frontend/src/metabase-lib/lib/metadata/Field.ts

```diff
@@ -245,6 +245,8 @@ class FieldInner extends Base {
     if (Array.isArray(this.id)) {
       // if ID is an array, it's a MBQL field reference, typically "field"
       return this.id;
+    } else if (this.field_ref) {
+      return this.field_ref;
```

Comment:
`dimension.isEqual(otherDimension)` was failing because one of the dimension was created via a `field.dimension()` call, which was not passing the options arg from its `field_ref` because Field's `reference` method (seen here) was not relying on its `field_ref` property, which contains the options in entry 2 of the field_ref (e.g. `["field", "COLUMN_NAME", { "base-type": "type/Foo" }]`)